msal/token_cache.py

```diff
@@ -136,20 +143,20 @@ def search(self, credential_type, target=None, query=None):  # O(n) generator
             preferred_result = self._get_access_token(
                 query["home_account_id"], query["environment"],
                 query["client_id"], query["realm"], target)
-            if preferred_result:
+            if preferred_result and self._is_matching(
+                preferred_result, query, target_set,
+            ):
```

Comment:
Under what circumstances can `_is_matching` return a `False`?

Reply:
The `self._get_access_token()` uses token key to find a token in `O(1)` time complexity. However, previously we forgot to also use `self._is_matching()` to double check the other criteria such as scopes and - in this bugfix's case - the `key_id` claim. This PR refactors the implementation to call `self._is_matching()` in both the `O(1)` code path and the traditional `O(N)` code path.

Reply:
I see. To make it concrete, when retrieving an SSH cert with 

```
target=['https://pas.windows.net/CheckMyAccess/Linux/.default']
```

`query` is

```py
{
    'client_id': 'a7003d8c-e50f-4371-91a6-ef37bba4ab23', 
    'environment': 'login.microsoftonline.com', 
    'home_account_id': None, 
    'key_id': '38b8ea33acd3d9dbc8eb0d9376466b2f80154c5bd2e771b39d31210b733cc466', 
    'realm': '54826b22-38d6-4fb2-bad9-b7b93a3e9c5a'
}
```

An access token's key only contains:

- `home_account_id`
- `environment`
- `client_id`
- `realm` (`tenant_id`)
- `target` (`scope`)

For example: `-login.microsoftonline.com-accesstoken-a7003d8c-e50f-4371-91a6-ef37bba4ab23-54826b22-38d6-4fb2-bad9-b7b93a3e9c5a-https://pas.windows.net/checkmyaccess/linux/.default`

As `query['key_id']` is not used as part of the key, an access token with different `key_id` is returned. Without checking the result again with `_is_matching`, the wrong access token is returned.